components/pd_client/src/util.rs

```diff
@@ -234,63 +237,42 @@ where
     Resp: Send + 'static,
     F: FnMut(&RwLock<Inner>, Req) -> PdFuture<Resp> + Send + 'static,
 {
-    fn reconnect_if_needed(mut self) -> Box<dyn Future<Item = Self, Error = Self> + Send> {
+    async fn reconnect_if_needed(&mut self) -> bool {
         debug!("reconnecting ..."; "remain" => self.reconnect_count);
 
         if self.request_sent < MAX_REQUEST_COUNT {
-            return Box::new(ok(self));
+            return false;
         }
 
         // Updating client.
         self.reconnect_count -= 1;
 
         // FIXME: should not block the core.
         debug!("(re)connecting PD client");
-        match block_on(self.client.reconnect()) {
+        match self.client.reconnect().await {
             Ok(_) => {
                 self.request_sent = 0;
-                Box::new(ok(self))
+                false
+            }
+            Err(_) => {
+                tokio::time::delay_for(Duration::from_secs(RECONNECT_INTERVAL_SEC)).await;
```

Comment:
AFAIK the timer of tokio 0.2 only works inside tokio's runtime and there's no way to set it up manually

Reply:
Yes, so is there any problem with this usage?

Reply:
The methods of `PdClient` don't necessarily run in the tokio runtime. For example, `block_on` is obviously not a tokio runtime.